core/storage/couchStorage.go

```diff
@@ -0,0 +1,1801 @@
+	result := couchObject{}
+	id := createObjectCollectionID(orgID, objectType, objectID)
+	if err := store.getOne(id, &result); err != nil {
+		return &Error{fmt.Sprintf("Failed to retrieve object. Error: %s.", err)}
+	}
+	for i, d := range result.Destinations {
+		d.Status = common.Delivering
+		result.Destinations[i] = d
+	}
+	result.LastUpdate = time.Now()
+	if err := store.upsertObject(id, result); err != nil {
+		return &Error{fmt.Sprintf("Failed to update object's destinations. Error: %s.", err)}
+	}
+	return nil
+}
+
+// StoreDestination stores the destination
+func (store *CouchStorage) StoreDestination(destination common.Destination) common.SyncServiceError {
+	id := getDestinationCollectionID(destination)
+	existingObject := &couchDestinationObject{}
+	newObject := couchDestinationObject{ID: id, Destination: destination}
+	if err := store.getOne(id, existingObject); err != nil {
+		if err != notFound {
+			return &Error{fmt.Sprintf("Failed to store a destination. Error: %s.", err)}
+		}
+		existingObject = nil
+	}
+	if existingObject != nil {
+		newObject.Rev = existingObject.Rev
+	}
+	err := store.upsertObject(id, newObject)
+	if err != nil {
+		return &Error{fmt.Sprintf("Failed to store a destination. Error: %s.", err)}
+	}
+	return nil
+}
+
+// RetrieveDestination retrieves a destination
+func (store *CouchStorage) RetrieveDestination(orgID string, destType string, destID string) (*common.Destination, common.SyncServiceError) {
+	result := couchDestinationObject{}
+	id := createDestinationCollectionID(orgID, destType, destID)
+	if err := store.getOne(id, &result); err != nil {
+		if err != notFound {
+			return nil, &Error{fmt.Sprintf("Failed to fetch the destination. Error: %s.", err)}
+		}
+		return nil, &NotFound{fmt.Sprintf(" The destination %s:%s does not exist", destType, destID)}
+	}
+	return &result.Destination, nil
+}
+
+// RetrieveDestinations returns all the destinations with the provided orgID and destType
+func (store *CouchStorage) RetrieveDestinations(orgID string, destType string) ([]common.Destination, common.SyncServiceError) {
+	result := []couchDestinationObject{}
+	var err error
+
+	if orgID == "" {
+		if destType == "" {
+			err = store.findAll(map[string]interface{}{"selector": map[string]interface{}{"last-ping-time": map[string]interface{}{"$exists": true}}}, &result)
+		} else {
+			err = store.findAll(map[string]interface{}{"selector": map[string]interface{}{"destination.destinationType": destType}}, &result)
+		}
+	} else {
+		if destType == "" {
+			err = store.findAll(map[string]interface{}{"selector": map[string]interface{}{"destination.destinationOrgID": orgID}}, &result)
+		} else {
+			query := map[string]interface{}{
+				"selector": map[string]interface{}{
+					"destination.destinationOrgID": orgID,
+					"destination.destinationType":  destType}}
+			err = store.findAll(query, &result)
+		}
+	}
+	if err != nil {
```

Comment:
if the error is not "NotFound", return error message, otherwise, return empty destination list. So line717 should be something like: